### PR TITLE
fix(evaluation): make spectral_matrix_sign scale-free via power-of-two frexp rescaling

### LINKED ISSUE #2379: bug(optimizer-geometry): the 1e-12 Frobenius floor launders small matrices into a numerically zero matrix sign
## Summary

`spectral_matrix_sign` divides by a fixed `1e-12` magnitude floor instead of the matrix's own Frobenius norm, so any matrix whose norm falls below that floor is **laundered into a numerically zero "valid" sign** rather than orthogonalized. The function's docstring promises the opposite:

> Frobenius normalization places every singular value in its convergence interval and preserves an exact zero for a zero matrix.

`alberta_framework/evaluation/optimizer_geometry.py:174-176`:

```python
norm = jnp.linalg.norm(value)
valid = jnp.all(jnp.isfinite(value)) & jnp.isfinite(norm)
x = value / jnp.maximum(norm, jnp.asarray(1e-12, dtype=value.dtype))
```

When `norm < 1e-12` the divisor is the floor, not the norm, so `‖x‖_F = ‖M‖_F / 1e-12 ≪ 1`. The NS5 recurrence `f(X) = 3/2 X − 1/2 XX^T X` is then applied to singular values far below its convergence interval: each step multiplies them by ≈1.5, so five steps return ≈`7.59 · σ / 1e-12` — a matrix that is numerically zero — while the transaction still reports `valid=True`.

In float32 the floor is reached earlier than the exponent range suggests, because the sum of squares underflows: for entries near `1e-20`, `jnp.linalg.norm` returns exactly `0.0` for a plainly nonzero matrix.

## Reproduction

```python
import jax
import jax.numpy as jnp
from alberta_framework.evaluation.optimizer_geometry import spectral_matrix_sign

m = jnp.array([[2.0, 1.0], [0.5, -1.0]], dtype=jnp.float32)
for scale in (1.0, 1e-13, 1e-15, 1e-20, 1e-25):
    tiny = m * jnp.asarray(scale, dtype=jnp.float32)
    print(scale, float(jnp.linalg.norm(tiny)), float(jnp.linalg.norm(spectral_matrix_sign(tiny))))
```

```
scale     norm(input)      norm(returned sign)   expected
1e+00     2.500000e+00     1.414202e+00          1.414202e+00
1e-13     2.500000e-13     1.201002e+00          1.414202e+00
1e-15     2.500000e-15     1.898316e-02          1.414202e+00
1e-20     0.000000e+00     1.898437e-07          1.414202e+00
1e-25     0.000000e+00     1.898438e-12          1.414202e+00
```

Every row is the same matrix up to a positive scalar, so every row must return the same sign. `valid` is `True` in all five.

Singular values of the returned "sign" (computed in float64):

| input scale | singular values | orthogonal? |
| --- | --- | --- |
| `1.0` | `[0.99999999, 0.99998346]` | yes |
| `1e-13` | `[0.96671583, 0.71264756]` | no |
| `1e-25` | `[1.69801409e-12, 8.49007043e-13]` | no — numerically zero |

The reverse-mode derivative is also lost: `jax.jacrev(spectral_matrix_sign)` is non-finite at `1e-25` scale and finite at unit scale.

## Consumer impact

`muon_ogd_dual_update_transaction` uses this sign as the update direction (`optimizer_geometry.py:260` and `:274`). On the documented empty-constraint path — pinned by `test_muon_ogd_empty_constraints_reduces_exactly_to_ns5` as reducing exactly to NS5 — the update vanishes with gradient scale instead of staying scale-free:

| gradient scale | `‖update‖_F` |
| --- | --- |
| `1.0` | `1.414202e+00` |
| `1e-13` | `1.201002e+00` |
| `1e-25` | `1.898438e-12` |

A scale-free orthogonalized direction is the entire point of the Muon-OGD step, so a small-magnitude momentum matrix silently produces no step rather than a unit-scale one.

The frozen streaming benchmark does not expose this: `_frozen_stream()` matrices have Frobenius norms between `1.15` and `2.59`, all far above the floor, which is why every published arm metric is unaffected.

## Why the floor is the wrong instrument

The existing `test_geometry_float32_overflow_is_invalid_not_laundered` pins the policy for the other end of the exponent range: when the input's own arithmetic overflows, the transaction must report `valid=False` rather than launder a plausible-looking result. The small end does the opposite — it returns a fabricated near-zero matrix with `valid=True`. A magnitude floor cannot distinguish "genuinely zero" from "small but well-conditioned"; only the norm can, and the norm is unusable in the input dtype at that magnitude.

## Proposed fix

Rescale by an exact power of two before normalizing, which is the pattern already used for the same hazard in `alberta_framework/utils/nexting.py`, `alberta_framework/core/oak.py`, `alberta_framework/core/state_builder.py`, and `alberta_framework/benchmarks/ipmnist_screening.py`:

```python
_, exponent = jnp.frexp(jnp.max(jnp.abs(value)))
rescaled = jnp.ldexp(value, -exponent)
rescaled_norm = jnp.linalg.norm(rescaled)
positive = rescaled_norm > 0
x = jnp.where(
    positive,
    rescaled / jnp.where(positive, rescaled_norm, jnp.ones_like(rescaled_norm)),
    jnp.zeros_like(rescaled),
)
```

The largest entry lands in `[0.5, 1)`, so the sum of squares can neither overflow nor underflow, and an exact zero is preserved by the zero test rather than by a magnitude floor. Because power-of-two scaling is exact, the quotient is bit-identical to `value / norm` for every input whose norm was already representable — verified bit-identical on 200 random float32 matrices spanning `2**-30` to `2**30`. `norm` stays in place as the overflow signal so the pinned invalid-on-overflow behaviour is unchanged.

## Environment

JAX 0.11.0, jaxlib 0.11.0, numpy 2.5.1, CPython 3.13.14, Windows 11 AMD64, `jax_enable_x64` False (default).

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2379

## Summary
In `alberta_framework/evaluation/optimizer_geometry.py`:
`spectral_matrix_sign_transaction` divided matrices by `jnp.maximum(norm, 1e-12)`. When a matrix norm fell below $10^{-12}$ (or entries dropped below $10^{-20}$ where sum-of-squares underflowed to 0.0), the divisor defaulted to $10^{-12}$ instead of the true Frobenius norm. As a result, singular values were placed far below the NS5 convergence interval, causing the recurrence to return a numerically zero matrix with `valid=True` rather than orthogonalizing the matrix.

## Proposed Changes
1. Rescaled input matrices by power-of-two magnitude normalization using `jnp.frexp` and `jnp.ldexp`.
2. Normalized the rescaled matrix by its rescaled Frobenius norm, preserving exact zero on genuinely zero inputs while eliminating the $10^{-12}$ floor bug.
3. Added unit tests in `tests/test_optimizer_geometry.py` verifying numerical equivalence of matrix signs across small input scales from $10^{-13}$ down to $10^{-30}$.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_optimizer_geometry.py` (29/29 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/evaluation/optimizer_geometry.py tests/test_optimizer_geometry.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/evaluation/optimizer_geometry.py` (clean).

### CODE REVIEW

**Review by lumix5 — CHANGES_REQUESTED** [review]
Independent verification at the current head (no prior review existed at any head).

**Disposition: fix requested — this rescale is the stronger half of the fix, but it must keep the all-subnormal fail-closed guard from #2476 to close the same hole that PR closes.**

What I measured in a throwaway container (network denied, branch module overlaid on unchanged main):
- **Tiny-normal matrices: this PR is better than the already-approved #2476.** At scale 1e-30 (entries normal, squares underflow), #2476 fails closed (valid=False) while this branch rescales by `ldexp(-frexp(max))` and returns the **true** sign matching the unscaled reference — I reproduced that directly.
- **All-subnormal matrices: the original hole is still open here.** For entries at 2.938736e-39 I measured `valid=True` with an all-zero output — `max_val` and `rescaled_norm` are both flushed to zero, so `positive=False` routes `x` to zeros, but `valid` never consults that condition (it only checks finiteness), so Newton runs on a zero matrix and the transaction is reported valid. #2476's bit-level `underflowed_nonzero = (norm == 0) & has_nonzero_bits` is exactly the missing predicate.

Requested change (small): after the rescale, extend `valid` with the nonzero-bits check — e.g. `valid = ... & ~((rescaled_norm == 0.0) & _subnormal_entry_present(value))` using the same bitcast-mask technique #2476 already added to this module — and add the all-subnormal regression test from #2476 to this branch. With that, this PR strictly dominates #2476 (true signs wherever recoverable, fail-closed only where flush-to-zero makes the answer unrecoverable), and #2476 can close in its favor.

Note for maintainers: #2476 (approved) and this PR edit the same function with complementary halves; whichever survives should carry both.

**Comment by lalalune** [comment]
Superseded by merged #2380. Reviewed this matrix-sign rescaling diff: exact power-of-two normalization now replaces the fixed 1e-12 floor. The selected implementation has 17 passing regressions covering scale invariance, JIT, gradients, float16/bfloat16, zero input, and the Muon-OGD consumer, with required CI green. Keeping another equivalent normalization branch would add no useful behavior.
